### PR TITLE
Overhauling the overhaul

### PR DESCRIPTION
**Pull request by daryltucker**
Changes
===

> There are a few significant and breaking changes.

Let's discuss, but these should help keep things maintainable.  Please let me know if you'd prefer that I pull anything out.

**Breaking**

1. `gist_config.yml` => `.gifc_config.yml`
1. Argument adjustments

**Other**

1. README Updates!
1. Re-worked for modular architecture
    * We can `python -m gifc` or `gifc` now!
    * `__main__.py` controls what runs when evoked
    * gifc.py has lower-level functions to allow extension
    * `setup.py`
    * TODO: Discuss versioning
1. Configuration file is now `.gifc_config.yml` from `gist_config.yml`
1. Configuration file can be found outside of binary/script directory
1. Unified argument language:
    * Update now uses `-fn` instead of `-f`, to be in-line with new flags
        * `-f` always references a local file
        * `-fn` always references one of the filenames of a file in the Gist, relative to GitHub
    * Update and Create now use all same flags
    * Create and Update share `-d` argument for both new and changes to Description
    * `--describe` -> `--description`
1. Added 'global' arguments
    * All 'actions' support:
        * `--debug`
        * `--editor EDITOR`
        * `--private`
1. `--editor` flag no longer checks against editor list.  Trust the User to know which editor to use
    * We now read $EDITOR bash variable, by default
    * `--editor` flag is honored above all else
    * The default/fallback editor, when no `$EDITOR` and no `--editor` is `nano` ( ;_; )
1. Improved help text and setup
1. arrow added as a soft dependency
    * TODO Discuss full dependency
1. Implemented three output functions:
    * printDebug(msg)
    * printError(msg, sysexit=True)
    * printTerminal(msg)
1. Colors -> Tmods class
    * Tmods.modifiers holds name to string mapping
    * Tmods.wrap(text, premod, postmod) to auto-wrap text
1. Removed `-p` flag, and forcing `--public` for clarity of such a critical option.
    * Both Public and Private begin with *P*.  Let's clarify by forcing users to declare `--public` explicitly.
1. Streamlined mapping between JSON response and format/output function

Testing
===

**Actions**

> Create a new Gist

```bash
$ gifc create -fn "Test_Remote.mkd" -f /tmp/Test.mkd -d "A test description"
Unable to import arrow: No module named 'arrow'
Gist 530c3f7c4e2b6c938a34ebc87abc3b9f successfully created
```

> Create a new Gist (DEBUG)

```bash

$ gifc create -fn "Test_Remote.mkd" -f /tmp/Test.mkd -d "A test description" --debug
Unable to import arrow: No module named 'arrow'
DEBUG: DEBUG Enabled.
DEBUG: Selected nano as editor.
DEBUG: Found configuration file: /home/daryl/.config/.gifc_config.yml
DEBUG: Found Config Variable w. length of 11 for USER_ID
DEBUG: Found configuration file: /home/daryl/.config/.gifc_config.yml
DEBUG: Found Config Variable w. length of 40 for TOKEN
DEBUG: Created http session w. authorization details.
DEBUG: Using credentials daryltucker:11111... against https://api.github.com.
DEBUG: Poping keys: ['public']
DEBUG: Generated Session: <bound method Session.post of <requests.sessions.Session object at 0x7fb588f4db70>>
DEBUG: Generated URL: https://api.github.com/gists
DEBUG: Generated Payload: {'description': 'A test description', 'files': {'Test_Remote.mkd': {'content': 'Test Data\n'}}}
Gist ffebee0abdc1cc25ccbd806004d23002 successfully created
```

> Update an existing Gist

```bash
$ gifc update ffebee0abdc1cc25ccbd806004d23002 -fn "Test_Remote.mkd" -f /tmp/Test.mkd -d "A new description" # or without -d to keep existing description
Unable to import arrow: No module named 'arrow'
Gist ffebee0abdc1cc25ccbd806004d23002 successfully updated
```

> Delete an existing Gist (DEBUG)

```bash
$ gifc delete ffebee0abdc1cc25ccbd806004d23002 --debug
Unable to import arrow: No module named 'arrow'
DEBUG: DEBUG Enabled.
DEBUG: Selected nano as editor.
DEBUG: Found configuration file: /home/daryl/.config/.gifc_config.yml
DEBUG: Found Config Variable w. length of 11 for USER_ID
DEBUG: Found configuration file: /home/daryl/.config/.gifc_config.yml
DEBUG: Found Config Variable w. length of 40 for TOKEN
DEBUG: Created http session w. authorization details.
DEBUG: Using credentials daryltucker:11111... against https://api.github.com.
Deleting ffebee0abdc1cc25ccbd806004d23002 w. 0 file(s): []!
DEBUG: Poping keys: []
DEBUG: Generated Session: <bound method Session.delete of <requests.sessions.Session object at 0x7f64d2aa72e8>>
DEBUG: Generated URL: https://api.github.com/gists/ffebee0abdc1cc25ccbd806004d23002
DEBUG: Generated Payload: {}
Gist ffebee0abdc1cc25ccbd806004d23002 deleted!
```

> Show a single Gist without arrow

```bash
$ gifc get 1
Unable to import arrow: No module named 'arrow'
Retrieving 1 Gists...
--------------------------------------------------------------------------------------------------------------------------------<List of Gists>---------------------------------------------------------------------------------------------------------------------------------
Description:            A test description
Updated:                2018-12-01T09:41:32Z (some time ago)
Created:                2018-12-01T09:41:32Z (some time ago)
Gist ID::               530c3f7c4e2b6c938a34ebc87abc3b9f
File List:              ['Test_Remote.mkd']
Web URL:                https://gist.github.com/530c3f7c4e2b6c938a34ebc87abc3b9f
---------------------------------------------------------------------------------------------------------------------------------------<>---------------------------------------------------------------------------------------------------------------------------------------
```

> Show a single Gist with arrow (`pip3 install arrow`)

```bash
$ gifc get 1
Retrieving 1 Gists...
--------------------------------------------------------------------------------------------------------------------------------<List of Gists>---------------------------------------------------------------------------------------------------------------------------------
Description:            A test description
Updated:                2018-12-01T09:41:32Z (5 minutes ago)
Created:                2018-12-01T09:41:32Z (5 minutes ago)
Gist ID::               530c3f7c4e2b6c938a34ebc87abc3b9f
File List:              ['Test_Remote.mkd']
Web URL:                https://gist.github.com/530c3f7c4e2b6c938a34ebc87abc3b9f
---------------------------------------------------------------------------------------------------------------------------------------<>---------------------------------------------------------------------------------------------------------------------------------------
```

------
**Help**

> Top-Level Help

```bash
$ gifc -h
usage: gifc [-h] {get,create,update,delete,remove} ...

Interact with GitHub Gists via command-line interface.

positional arguments:
  {get,create,update,delete,remove}
    get                 Retrieve Gists from GitHub.
    create              Create a new Gist on GitHub.
    update              Update an existing Gist.
    delete              Delete a Gist. DANGER! No recovery!
    remove              Remove individual files from a specific Gist.

optional arguments:
  -h, --help            show this help message and exit
```

> Action-specific Help

```bash
$ gifc create -h
usage: gifc create [-h] [--debug] [-e EDITOR] [-p] -fn GH_FILE
                   [-d DESCRIPTION] [-m MESSAGE | -f FILE | -i]

optional arguments:
  -h, --help            show this help message and exit
  --debug               Enable debug output.
  -e EDITOR, --editor EDITOR
                        Use an editor other than bash default $EDITOR.
  -p, --public          Make a public Gist. Omit for default, private Gist.
  -fn GH_FILE, --gh_file GH_FILE
                        Remote filename of Gist.
  -d DESCRIPTION, --description DESCRIPTION
                        Description of Gist.
  -m MESSAGE, --message MESSAGE
                        Gist contents from CLI as string.
  -f FILE, --file FILE  Gist contents from a local file
  -i, --interactive     Use editor to define Gist after save/close. See
                        --editor. Default $EDITOR.
```

### CODE REVIEW

**Review by armsp — CHANGES_REQUESTED** [review]
Hello Daryl. I am new to Pull Requests. I have commented somewhere extensively about what i like and what I feel reserved about. I would just like to add one more thing, I cannot accept any code that does NOT follow PEP-8. I can edit your code myself and push those changes myself, but I am not sure if you'd want that as I don't know if it will reflect in your contributions. I am figuring that out. In case you want to read what I had read elsewhere, read the following -  

This is brilliant work @daryltucker . However I must point out a few things -  
This is the first time someone made a pull request for my project. I am not familiar with how to handle that. It will take some time for me to review and address your merge requests properly.  

Lets talk about the changes you submitted -  
### Changes I like

1. Adding `.gitignore`
2. Using **Etymology** instead of **Name**
3. I really appreciate the extensive documentation you have written in a properly divided way. I think I will incorporate that albeit with some changes.
4. TMod
5. Even I wanted to find out programmatically the terminal size for printing properly but couldn't figure out why. Have not tested it as of now. but if it works I am definitely going to add it.

### Changes I would not want to make

* I am not in favor of all the places you have used Quotes. I would request you to leave that to me.
* Take a look at the installation instructions for the alpha release. I think it much more succinct. We don't have to worry about changes, tests etc. Just pull from remote and the changes are reflected. It's better if anyone who wants to use this doesn't have to do much more. Otherwise they just leave it. I don't want that. I want to make it a very simple process. Let me know if you feel any different.
* Updating gists do not require both file name and gist id. It was a mistake I made. Refer to the current alpha release. It has been corrected. Let me know if it doesn't work the way I think it works.
* Its better to have the LICENSE in the root directory. 
* I must point out that we cannot use all the available editors. Editors like nano, vim, gedit block the subprocess which is perfect. However if you open sublime, VSCode in subprocess they detach themselves and the program terminates without your changes. Please refer to this [SO post](https://stackoverflow.com/questions/53288147/calling-vs-code-from-subprocess-is-non-blocking-unlike-other-editors) I raised for that. Unfortunately no help.

### What I am unsure of i.e things I may have to discuss with other potential contributors or time !

1. **Using Arrow**   
The point is, I do not think that information about dates would be of much use. First of all let us think how people use gists. They are almost always in a hurry, like they want to use it to embed in their Medium article or some code snippet that they find useful or something quick and other quirky ways like TODO lists, private codes etc.  
In that case showing them additional information would just end up taking more screen space. I want something neat and clean and trimmed to the most relevant information.   
Maybe we can release it as an additional optional thing later on. Lets skip it altogether for now. **Feel free to make another pull request later on after I make the first beta release.**

2. Using yaml file for configuration management.
I do not know what is the best way to manage those git credentials. I used yaml files earlier, but as you might have seen for the first alpha release, i have removed it and am using environment variables instead. I DO NOT know the security issues if any. I am researching on it and would like to converge on something concrete. It is open for debate.

3. The setup file that I have now uses `scripts` instead of `entry_points`. I believe the former is much simpler and is less worrisome as all it does is just add this script to PATH so that it becomes a callable. I am not sure why i did not use entry_points when i was researching about setup tools. I guess I need more clarity to decide this one. But I know for sure that I want something that has the least footprint on the system so that those that end up uninstalling `gifc` don't have to worry about leftover artefacts.

4. Your division of labor (Delete, Upsert, Retrieve) is pretty nice. But I need some time to decide if i want to represent the process that way.

I am yet to go through your code base. Need some time as I was traveling.

**Review by armsp — COMMENTED** [review]
Okay, so I have gone through how to handle pull requests and I will start reviewing your code and commenting an the specific lines soon. I have one request though. You did follow almost all the rules, however it would be much easier for me to merge small parts of your work. So please don't commit such a huge overhaul at once to git. Keep committing locally and eventually push them to remote. So that when you make a pull request, it comprises of many commits with small changes that is easier to review and change and eventually incorporate to master.

**Comment by daryltucker** [comment]
Hey! Thanks so much for the response.  I am still going through them but I appreciate your sharing these details with me.

I make attempts to follow pep8 styling.  I don't usually run `pyflakes` and the others, as I find them to be ridiculous and make the code less readable.  No problems making adjustments as you see fit.

The LICENSE file was placed within the Module root.  We have to think of the repository directory similar to a web root, and the Module directory, which is build and installed with setup.py, should also have a LICENSE.  I have not looked into this exactly, but I believe that the symlink here should provide GitHub with a license, and the python package with a license, but we'll have to double check this when we package, and if you decide to publish to a module repo.

I agree with you on your view of `arrow`, and is why I made the shiv.  I can remove the shiv and all `arrow` code, absolutely.  I couldn't think of any additional use cases, but I wanted to see if you could think of any.

I'll get working on these adjustments and if you could make comments against any line numbers, etc, I'll dig into those comments as well.

Again, I appreciate your willingness to discuss these changes.

**Comment by daryltucker** [comment]
When publishing changes, I ensure to squash the changes into a single, relative changeset.  When I saw that you made some overhaul changes, I thought this might be a good time to slip in a large changeset.

That being said, I can pull things out and provide another clean changeset, but many of these changes are interdependent, and I believe that each commit should really be a valid or working step towards another goal, vs arbitrarily breaking out into multiple commits.  I hope that this makes sense :D

**Comment by daryltucker** [comment]
> Using yaml file for configuration management.
I do not know what is the best way to manage those git credentials. I used yaml files earlier, but as you might have seen for the first alpha release, i have removed it and am using environment variables instead. I DO NOT know the security issues if any. I am researching on it and would like to converge on something concrete. It is open for debate.

This is a good question and there will be varying suggestions based on threat.  Here, we look in common configuration directories, and other common directories such as the current working directory, for the configuration file.

When using OAuth, we are able to create a credential for a single application, such as `gifc`, and limit the access or permissions that that credential has.  This helps protect the account.   We can educate, or guide Users to only allowing Gist-related permissions when they generate the token.

When we store the credentials, often they are stored in a configuration file, in clear text, and it is up to the systems file permissions to ensure that only the proper users or groups may access the file.  You can look into examples, such as Ansible, AWS, etc.  This is common on *nix systems.  This also means that multiple users can use the same binary, but each have their own set of credentials.

You can further obfuscate by allowing the configuration to be provided as a bash variable, exactly how we adjusted for `$EDITOR`.  This allows no file to exist, but the system to create an env variable, say `GIFC_GH_TOKEN`, and we would consume that first, and only look for the credentials in a file if we don't have the credentials in the environment.

We could consider `.gifc`, as it would be closer to standards.  This is an important decision that we can't keep changing so I'll stand by and wait for a decision on that one.

The changes surrounding the `.gifc_config` are good steps in the right direction to get us to these and other goals.

**Comment by armsp** [comment]
Hi @daryltucker 
I have made a **development** branch. Lets work on that branch from now on.   
I have also added a CONTRIBUTING.md file.  
I realized that its not possible to merge and approve such a large overhaul and I am still not sure how will I manage the merge conflicts. It would become a very time consuming process. So I will keep this pull request as it is for historical and referring purposes.  

I have found out items that won't break the existing code base. Hence I request you to make a (or multiple) pull request(s) to development branch and push those changes only. I am yet to review your code and put comments. I will do that in a while.  
Meanwhile I have tested the `term_width` function that you wrote and it seems to work just fine. So you can make a first pull request for that functionality to the **development** branch. Hope you don't mind the additional work.  
I would also like to add you as contributor while committing to _README.md_. For that I need your email so that I can add you next time I commit to _README.md_.  

Follow this thread for more items that don't break the code base and that you can make a pull request for.

Cheers!!

File: gifc/gifc.py
Comment (armsp) [inline gifc/gifc.py:61]: I think we can make this function a part of the `TerminalModifier` class. Maybe as a `classmethod` because it doesn't actually use any class information, but logically belongs to that class. For the new pull request as I suggested earlier in the comment, maybe we can ignore the specific `printDebug`, `printTerminal` functions and just use the regular `print`. We can add them later as and when needed.

File: gifc/gifc.py
Comment (armsp) [inline gifc/gifc.py:63]: Do we have any reason for using 80 as a fallback value or is it random?

File: gifc/gifc.py
Comment (armsp) [inline gifc/gifc.py:65]: I would want you to use `os.get_terminal_size(fd=1)` explicitly. (file descriptors 0, 1 and 2: stdin, stdout and stderr)
```suggestion
        width = os.get_terminal_size(fd=1)[0]
```

File: gifc/gifc.py
Comment (armsp) [inline gifc/gifc.py:81]: I don't understand why do we need an empty key: value pair ?

File: gifc/gifc.py
Comment (armsp) [inline gifc/gifc.py:102]: Perhaps `print` as a function name is not quite appropriate since it clashes with the standard librarie's print method.

**Review by armsp — CHANGES_REQUESTED** [review]

File: gifc/gifc.py
Comment (daryltucker) [reply gifc/gifc.py:63]: 80 characters is the standard terminal width.  It's also part of those `pep` guidelines :D.  

Please let me know if you have any other suggestion, but a default value is best for error prevention.

File: gifc/gifc.py
Comment (daryltucker) [reply gifc/gifc.py:65]: The terminal width should be the same for all, but... sure you're not wrong.  My concern would be that everything was tested without the file descriptor being selected, but doesn't see like a hard thing to test :D

File: gifc/gifc.py
Comment (daryltucker) [reply gifc/gifc.py:81]: For none, or null strings.  It's to prevent errors when interacting with this under certain logic paths.  Please let me know if I should go into further depth, but we want to make sure if someone selects 'no modifier', no modifier is used.

File: gifc/gifc.py
Comment (daryltucker) [reply gifc/gifc.py:102]: This is a method of the class, and was the clearest option.  I'm not concerned with this, but if you provide another name, it's not a big deal to change this method name, regardless of clash or  non-clash.

**Comment by armsp** [comment]
@daryltucker could you please send me an empty email (from the mail private or otherwise, connected to your github account) to my address so that I can add you while committing some changes to `README` and `gifc` on the master branch. I am thinking of creating another branch for you and merging your pull request there. The good changes there would be merged to development branch which would eventually be merged to master after I feel that it follows the vision that I have for gifc.  
Let me know if you think otherwise.

File: gifc/gifc.py
Comment (armsp) [reply gifc/gifc.py:102]: Cool

File: gifc/gifc.py
Comment (armsp) [reply gifc/gifc.py:65]: Yeah, I have already tested it and the thing is by default one of those file descriptors is being sent anyways. I just wanted to make it **explicit** :wink:

File: gifc/gifc.py
Comment (armsp) [reply gifc/gifc.py:63]: Not quite true actually. PEP-8 suggests you to keep 79 characters per line while coding. It has got nothing to do with what size the terminal is and how it displays there.   
On the other hand, we anyways need a default value. I am cool with 80. No issues on my part.

**Comment by armsp** [comment]
@daryltucker could you please make the same pull request in the development branch. Merging this to master would break it and I wan't the master to stay as it is. We can collaborate on development and then eventually I can merge it with master.

**Pull request merged**
